articles/python-generators/index.md

```diff
@@ -0,0 +1,160 @@
+
+To access the values produced by the generator expression, you need to call `next()` on it, just like you would with any other iterator. When there isn't next value in the generator object, It throws a `StopIteration` exception. 
+
+```python
+>>> next(generator_expression)
+'Hello world'
+>>> next(generator_expression)
+'Hello world'
+>>> next(generator_expression)
+'Hello world'
+>>> next(generator_expression)
+StopIteration
+```
+A for loop can be used to iterate the generator object.
+
+```python
+for expression in generator_expression:
+    print(expression)
+```
+
+
+
+Given a range of n numbers to filter out odd numbers. We could solve the problem using three different methods.
+1. Using function.
+
+```python
+def even_integers_function(n):
+    result = []
+    for i in range(n):
+        if i%2==0:
+            result.append(i)
+    return result
+```
+
+```python
+>>>even_integers_function(10)
+[0,2,4,6,8,10]
+```
+
+2. Using Generator function. 
+   The generator function below yields a generator object which we can iterate through to get the list of even numbers. 
+
+```python
+def even_integer_generator(n):
+    for i in range(n):
+        if i % 2 == 0:
+            yield i
+```
+A generator function returns a generator object which can be iterated to get the values.
+```python
+>>>even_integer_generator(10)
+<generator object even_integer_generator at 0x1036c3200>
+```
+
+```python
+>>>list(even_integer_generator(10))
+[0,2,4,6,8,10]
+```
+3. Using Generator expression.
+```python
+even_integers = (i for i in range(n) if i%2==0)
+```
+Given a list of names to convert each name in the list to uppercase.
+```python
+name_list = ["Adam","Eve","John","Doe","Peter","Paul","Kevin"]
+```
+1. User of functions.
+```python
+def names_to_uppercase(names):
+    uppercase_names = []
+    for name in names:
+        name.upper()
+        uppercase_names.append(name)
+    return uppercase_names
+```
+2. Use of generator functions
+```python
+def names_to_uppercase_generator(names):
+    for name in names:
+        name.upper()
+        yield name
+```
+3. User of generator expressions.
+```python
+uppercase_names = (name.upper() for name in name_list)
+```
+
+**Generating Fibonacci sequence using generator functions.**
+The Fibonacci Sequence is the series of numbers where next number is found by adding up the two numbers before it.
+0, 1, 1, 2, 3, 5, 8, 13, 21, 34, ...
+```python
+def fibonacci_gen():
+    trailing,leading = 0,1
+    while True:
+        yield leading
+        trailing,leading = leading, trailing+leading
+```
+To get the first value of the Fibonacci sequence we generated using generator function above we call the `fibonacci_gen()` function and assign it to variable fib. We then call `next()` on the fib variable, `fib.next()` to get the next number in the sequence.
```

Comment:
When referring to a variable in your explanation, use the \` \` symbol. for example, variable `fib`